Add nicer assertions to SiriTimetableSnapshotSourceTest

habrahamsson-skanetrafiken opened the pull request:
### Summary

Inspired by the raptor module tests I wanted to make the assertions easier to work with in SiriTimetableSnapshotSourceTest.

The previous assertions were pretty verbose since you had to assert each call by itself. Now you can write something along the lines:

```
  assertEquals(
      "UPDATED | A1 0:00:15 0:00:15 | B1 0:00:25 0:00:25",
      env.getRealtimeTimetable(env.trip1)
  );
```

instead of this:

```
    var tripTimes = env.getTripTimesForTrip(env.trip1);
    assertEquals(RealTimeState.UPDATED, tripTimes.getRealTimeState());
    assertEquals(11, tripTimes.getArrivalTime(0));
    assertEquals(15, tripTimes.getDepartureTime(0));
    assertEquals(20, tripTimes.getDepartureTime(1));
    assertEquals(25, tripTimes.getArrivalTime(1));
```

There are a lot of open PRs touching this piece of code so there might be some conflicts. But since this makes it easier to write these tests I thought it might be worth creating this PR anyway.

### Documentation

Added a little bit of documentation

vpaturet commented:
I have just realized that expected and actual are inverted in assertFailure(), maybe you can apply the fix in this PR?

leonardehrenfried commented:
I think this is a nice addition but in the GTFS case we are going to need a realtime state for each each stop. Something like: scheduled, updated, cancelled/skipped, no data. In Siri there is also "prediction inaccurate" and "recorded".

Would you want to add something like that or should that be GTFS-specific?

habrahamsson-skanetrafiken commented:
> I think this is a nice addition but in the GTFS case we are going to need a realtime state for each each stop. Something like: scheduled, updated, cancelled/skipped, no data. In Siri there is also "prediction inaccurate" and "recorded".
> 
> Would you want to add something like that or should that be GTFS-specific?

Sure, we should probably add the stuff that is generally interesting for most tests. And if there are some specific attributes that are only set in rare cases those can be manually asserted.

If we want to add realtime state for each stop, perhaps it's best to use abbreviations to not make it to verbose? Something like:

```
UPDATED | A1 [C] 0:00:15 0:00:15 | B1 [ND] 0:00:25 0:00:25 | C1 [U] 0:00:25 0:00:25"
```
C: Canceled
U: Updated
ND: No data
PI: Prediction inaccurate
R: Recorded

Or something along those lines?

leonardehrenfried commented:
Looks great.

leonardehrenfried reviewed (approved):
This looks great. I will be using it in the GTFS tests as well.

vpaturet reviewed (approved):
Looks good!

t2gran commented on src/ext-test/java/org/opentripplanner/ext/siri/SiriTimetableSnapshotSourceTest.java:
```diff
@@ -565,6 +549,80 @@ public TripTimes getTripTimesForTrip(Trip trip) {
+     * ND: No Data
+     * </pre>
+     */
+    private String encodeTimetable(TripTimes tripTimes, TripPattern pattern) {
```
I like this, maybe it can be moved into the core code with TripTimes and TripPattern. For example in a utility class like `TripTimesStringBuilder`  (like the PathStringBuilder) - then we can reuse it across tests an if we want also use i in debugging/logging. If you want you can put it in the test source tree, and we can move it to main if needed later. I think it is usful in both test and main - it is up to you.

habrahamsson-skanetrafiken replied:
Good idea! I moved it into main.
